web/src/copy_and_paste.ts

```diff
@@ -661,15 +699,33 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
 
         // Only intervene to generate formatted links when dealing
         // with a URL and a URL-safe range selection.
-        if (isUrl(trimmed_paste_text) && is_safe_url_paste_target($textarea)) {
-            event.preventDefault();
-            event.stopPropagation();
-            const url = trimmed_paste_text;
-            compose_ui.format_text($textarea, "linked", url);
-            return;
+        if (isUrl(trimmed_paste_text)) {
+            if (is_safe_url_paste_target($textarea)) {
+                event.preventDefault();
+                event.stopPropagation();
+                const url = trimmed_paste_text;
+                compose_ui.format_text($textarea, "linked", url);
+                return;
+            }
+            if (!compose_ui.cursor_inside_code_block($textarea) && !compose_ui.shift_pressed) {
+                event.preventDefault();
+                event.stopPropagation();
+                // Try to transform the url text to #**stream>topic** syntax
+                // if it is a valid url.
+                if (!cursor_at_markdown_link_marker($textarea)) {
+                    const syntax_text = try_stream_topic_syntax_text(trimmed_paste_text);
+                    if (syntax_text) {
+                        // It is a nice touch to get back the actual pasted url on undoing.
+                        add_text_and_select(trimmed_paste_text, $textarea);
+                        compose_ui.insert_and_scroll_into_view(syntax_text + " ", $textarea);
+                        return;
+                    }
+                }
+                compose_ui.insert_and_scroll_into_view(trimmed_paste_text, $textarea);
+                return;
```

Comment:
Why is it correct to call `compose_ui.insert_and_scroll_into_view` in this fallthrough case? 

I guess we know the text is just a URL, so the transformations that might happen below would be noops; but an alternative theory, for example, would be to not do `event.preventDefault/event.stopPropagation` in this case and just return from this function, allowing the native browser behavior to do the paste. 

How did you decide which option to select here? It may not matter, but I'd like to see reasoning, which we can then document in a comment above this line.

Reply:
There is no particular reason for that apart from having full control over what is pasted. 
I've switched to the alternative approach, since it simplifies the conditionals further.

Reply:
There is no particular reason for that apart from having full control over what is pasted.
I've switched to the alternative approach, since it simplifies the conditionals further.